### ISSUE TITLE
Chapter 10 -> Exercise 10 -> Question 10

### ISSUE THREAD

**Issue by kasrahabib** [ageron/handson-ml2#188]
Hi @ageron,

Thanks for the 2nd edition, it is an amazing book (so far I enjoy reading it as I am on a vacation).

I am struggling with exercise 10 (Chapter 10) question 10: 
The question is: "Train a deep MLP on the MNIST dataset ........ See if you can get over 98% precision. Try searching for the optimal learning rate .... i.e., by growing the learning rate exponentially, plotting the error, and finding the point where the error shoots up. ”

I did implement a callback to grow the learning rate and plotted the graph as it was asked. I end up with a smoother [2] line than a squiggly line (which is in your solution see[1]), then I hover over to your notebook (10_neural_nets_with_keras.ipynb) and re-run what is basically written there (with no changes) and end up with the same graph (a smoother one):

[1]. screenshot from your solution:
<img width="340" alt="Screen Shot 2020-06-19 at 6 47 25 PM" src="https://user-images.githubusercontent.com/29026528/85142287-5a9a8480-b25d-11ea-9587-4bfc3f66eb4b.png">

[2] the graph which got generated after I run your notebook on my computer:
![Unknown-1](https://user-images.githubusercontent.com/29026528/85141191-a1877a80-b25b-11ea-93fc-4f0fa13da2ce.png)

by the way, the val_loss and loss get NaN in the few last iterations:
1719/1719 [==============================] - 5s 3ms/step - loss: nan - accuracy: 0.6030 - val_loss: nan - val_accuracy: 0.0958

According to this (smoother graph), it seems that the loss shoots up at (exactly) 8e-1 and I set the learning rate to be (first 7e-1, next 6e-1 and then 5e-1) but no chances of getting around 97. and 96. %. 

Basically, my question (sorry if it might be very simple) is what is wrong? 
1- Must always this graph be a squiggly one like yours?
2- Is this (smoother) graph [2] a problem? If it is not then, is setting lr = 7e-1 a good choice?

Thanks in advance,
Kasra

**Comment by Devanshchowdhury2212** [comment]
Same problem here

**Comment by Praful932** [comment]
@kasrahabib @Devanshchowdhury2212 It is mentioned 98% precision, shouldnt it be accuracy. Both are different terms right. Precision is for each class. Or Did I miss something?

**Comment by kasrahabib** [comment]
> @kasrahabib @Devanshchowdhury2212 It is mentioned 98% precision, shouldnt it be accuracy. Both are different terms right. Precision is for each class. Or Did I miss something?

Yes, there is difference between precision and accuracy; by accuracy meant that, however I believe here if you specify the model to return the accuracy it returns the precision.

**Comment by Praful932** [comment]
> > @kasrahabib @Devanshchowdhury2212 It is mentioned 98% precision, shouldnt it be accuracy. Both are different terms right. Precision is for each class. Or Did I miss something?
> 
> Yes, there is difference between precision and accuracy; by accuracy meant that, however I believe here if you specify the model to return the accuracy it returns the precision.

According to what I [read](https://stackoverflow.com/questions/41531695/how-does-keras-define-accuracy-and-loss), Tensorflow figures on the go what the the accuracy metric should be according to the given loss function. If we need precision/recall we can [define](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/) it with the `metrics` that we give.

**Comment by kasrahabib** [comment]
> > > @kasrahabib @Devanshchowdhury2212 It is mentioned 98% precision, shouldnt it be accuracy. Both are different terms right. Precision is for each class. Or Did I miss something?
> 
> > 
> 
> > Yes, there is difference between precision and accuracy; by accuracy meant that, however I believe here if you specify the model to return the accuracy it returns the precision.
> 
> 
> 
> According to what I [read](https://stackoverflow.com/questions/41531695/how-does-keras-define-accuracy-and-loss), Tensorflow figures on the go what the the accuracy metric should be according to the given loss function. If we need precision/recall we can [define](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/) it with the `metrics` that we give.

Yep, I already have that one seen on the documentation. But what in the solution he did is; he return accuracy instead of precision and that makes me still believe that here that accuracy returns precision 😁

**Comment by Praful932** [comment]
> > > > @kasrahabib @Devanshchowdhury2212 It is mentioned 98% precision, shouldnt it be accuracy. Both are different terms right. Precision is for each class. Or Did I miss something?
> > 
> > 
> > > 
> > 
> > 
> > > Yes, there is difference between precision and accuracy; by accuracy meant that, however I believe here if you specify the model to return the accuracy it returns the precision.
> > 
> > 
> > According to what I [read](https://stackoverflow.com/questions/41531695/how-does-keras-define-accuracy-and-loss), Tensorflow figures on the go what the the accuracy metric should be according to the given loss function. If we need precision/recall we can [define](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/) it with the `metrics` that we give.
> 
> Yep, I already have that one seen on the documentation. But what in the solution he did is; he return accuracy instead of precision and that makes me still believe that here that accuracy returns precision

Yea even I thought that too, its just accuracy tho.
For the exercise the author has made a custom callback for each batch end, Did you use a similar approach too?
I used a Keras callback which increase the `learning rate` by a constant at the end of each epoch.
I got this graph using it and a final Test accuracy of 98.06.

![image](https://user-images.githubusercontent.com/45713796/95499650-3ce75b80-09c3-11eb-96a3-1275c72d34b1.png)
As for the author's graph, I think it is because of small batch size or something similar in a older version of keras, which caused the rapid up and down of losses.

**Comment by Rouizi** [comment]
I faced the same issue. I hope we can get an answer.
Also when rolling back to the best model after the training is finished I get a very poor accuracy:
```
my_model = keras.models.load_model("my_mnist_model.h5") # rollback to best model

my_model.evaluate(X_test, y_test)
313/313 [==============================] - 0s 1ms/step - loss: 0.0616 - accuracy: 0.0977
[0.061620958149433136, 0.09769999980926514]```
```
But without using the best model, just making evaluation I get 98%:
```
model.evaluate(X_test, y_test)
313/313 [==============================] - 0s 1ms/step - loss: 0.0751 - accuracy: 0.9845
[0.0750538781285286, 0.984499990940094]
```
There is something wrong with callbacks.

**Comment by ssuhas76** [comment]
> Hi @ageron,
> 
> Thanks for the 2nd edition, it is an amazing book (so far I enjoy reading it as I am on a vacation).
> 
> I am struggling with exercise 10 (Chapter 10) question 10:
> The question is: "Train a deep MLP on the MNIST dataset ........ See if you can get over 98% precision. Try searching for the optimal learning rate .... i.e., by growing the learning rate exponentially, plotting the error, and finding the point where the error shoots up. ”
> 
> I did implement a callback to grow the learning rate and plotted the graph as it was asked. I end up with a smoother [2] line than a squiggly line (which is in your solution see[1]), then I hover over to your notebook (10_neural_nets_with_keras.ipynb) and re-run what is basically written there (with no changes) and end up with the same graph (a smoother one):
> 
> [1]. screenshot from your solution:
> <img alt="Screen Shot 2020-06-19 at 6 47 25 PM" width="340" src="https://user-images.githubusercontent.com/29026528/85142287-5a9a8480-b25d-11ea-9587-4bfc3f66eb4b.png">
> 
> [2] the graph which got generated after I run your notebook on my computer:
> ![Unknown-1](https://user-images.githubusercontent.com/29026528/85141191-a1877a80-b25b-11ea-93fc-4f0fa13da2ce.png)
> 
> by the way, the val_loss and loss get NaN in the few last iterations:
> 1719/1719 [==============================] - 5s 3ms/step - loss: nan - accuracy: 0.6030 - val_loss: nan - val_accuracy: 0.0958
> 
> According to this (smoother graph), it seems that the loss shoots up at (exactly) 8e-1 and I set the learning rate to be (first 7e-1, next 6e-1 and then 5e-1) but no chances of getting around 97. and 96. %.
> 
> Basically, my question (sorry if it might be very simple) is what is wrong?
> 1- Must always this graph be a squiggly one like yours?
> 2- Is this (smoother) graph [2] a problem? If it is not then, is setting lr = 7e-1 a good choice?
> 
> Thanks in advance,
> Kasra

Hello Kasra,

I have just started with the chapter 10 and am training the NN for fashion_mnist. Do you know why the epoch size is coming as 1719/1719? I see from your question that you are also getting the same. Shouldn't it be as per the texbook shown as below?

Train on 55000 samples, validate on 5000 samples
Epoch 1/30
**55000/55000** [==============================] - 2s 44us/sample - loss: 0.7226 - accuracy: 0.7641 - val_loss: 0.5073 - val_accuracy: 0.8320

**Comment by kasrahabib** [comment]
Hi @ssuhas76,

I noticed that too, that is relative to the number of instance in your batches. Currently, the default number of instances assigned to a batch is 32 (try to change it and you will see that number is changing, or in oder words: training_set_size/batch_size = 1718.5 ~ 1719   <--- keras rounds it), and I think at the time the book's tutorial was written it was diffrent.

Cheers,
Kasra

**Comment by kasrahabib** [comment]
Hi @ssuhas76,

I noticed that too, that is relative to the number of instance in your batches. Currently, the default number of instances assigned to a batch is 32 (try to change it and you will see that number is changing, or in oder words: training_set_size/batch_size = 1718.5 ~ 1719   <--- keras rounds it), and I think at the time the book's tutorial was written it was diffrent.

Cheers,
Kasra

**Comment by kasrahabib** [comment]
Hi @ssuhas76,

I noticed that too, that is relative to the number of instance in your batches. Currently, the default number of instances assigned to a batch is 32 (try to change it and you will see that number is changing, or in oder words: training_set_size/batch_size = 1718.5 ~ 1719   <--- keras rounds it), and I think at the time the book's tutorial was written it was diffrent.

Cheers,
Kasra

**Comment by ageron** [comment]
Hi there,
Great question! There were a few (mostly minor) changes in TensorFlow since I wrote the book: during training, Keras now displays the batch number rather than than the sample number (batch number = sample number // batch size, as @kasrahabib mentioned). Also, it looks like `logs["loss"]` is now smoother, as you noticed. The change is not due to any change in my code since I wrote the book: indeed, you can actually run the [old code here](https://colab.research.google.com/github/ageron/handson-ml2/blob/2bd7ca14ab62e5531523b4796e5a8ba981b789cc/11_training_deep_neural_networks.ipynb), and you will see that the curve is now smooth. So it must be a change in TensorFlow. I'll investigate further.

**Comment by ageron** [comment]
Okay, got it: in TensorFlow 2.2.0, tf.keras made `logs["loss"]` return the mean loss since the start of the epoch, whereas it used to be the batch loss. Sadly, this was not documented in the [release notes](https://github.com/tensorflow/tensorflow/releases/tag/v2.2.0).

This means that what we are seeing in the new graph is actually the mean loss since the start of each epoch (in my code I just used one epoch, so we're really seeing the mean loss since the start of training). This is why the graph looks much smoother now. It's not necessarily a bad thing, since it removes most of the noise from the graph. However, it means that when the learning rate is too high and the batch loss starts exploding, it will take a few batches until it's clearly visible in the graph. In other words the graph is smoother AND shifted to the right (and since the loss starts high, the graph is also shifted up). As long as you know it, it's okay, you just need to account for that and pick a lower learning rate than you would have with the noisy graph.

I'll add a note in the notebook to explain this.

To figure this out I had to dig into TF Keras's source code, since the documentation does not explain what `logs["loss"]` actually is. I saw that it now uses `self.total_loss` when defining `logs["loss"]`. To be absolutely sure, I also ran the following code, which defines a custom MSE loss which prints out the batch loss, and a callback which prints out `logs["loss"]`. We can then compare the two. Using Colab, I installed TF 2.0, 2.1 and 2.2, and I compared the outputs: before 2.2, the `logs["loss"]` was equal to the batch loss, but starting at TF 2.2, it became the mean loss over the epoch.

```python
import numpy as np
import tensorflow as tf
from tensorflow import keras

X = np.zeros([4*3, 3])
y = np.zeros([4*3, 1])
y[:4] = 1000

class ShowBatchLoss(keras.callbacks.Callback):
    def on_batch_end(self, batch, logs):
        print("logs loss:", logs["loss"])

def my_loss(y_true, y_pred):
    squares = tf.square(y_pred - y_true)
    batch_loss = tf.reduce_mean(squares)
    tf.print("batch loss", batch_loss)
    return squares

model = keras.models.Sequential([keras.layers.Dense(1)])
model.compile(loss=my_loss, optimizer="sgd")
model.fit(X, y, epochs=10, batch_size=4, shuffle=False, callbacks=[ShowBatchLoss()])
```

Here's the output of this code in TF 2.0 and 2.1:

```
TF 2.1

Train on 12 samples
Epoch 1/10
batch loss 1e+06
logs loss: 1000000.0
 4/12 [=========>....................] - ETA: 3s - loss: 1000000.0000
batch loss 400
logs loss: 400.0
batch loss 384.16
logs loss: 384.16
12/12 [==============================] - 2s 133ms/sample - loss: 333594.7200
Epoch 2/10
batch loss 961952.938
logs loss: 961952.94
...
```

Notice that the batch loss and logs loss are equal (except for small rounding errors).

In TF 2.2, we get this:

```
Epoch 1/10
batch loss 1e+06
logs loss: 1000000.0
1/3 [=========>....................] - ETA: 0s - loss: 1000000.0000
batch loss 400
logs loss: 500200.0
batch loss 384.16
logs loss: 333594.71875
3/3 [==============================] - 0s 4ms/step - loss: 333594.7188
Epoch 2/10
batch loss 961952.938
logs loss: 961952.9375
...
```

The first two batch losses are 1e+06 and 400.
The first two logs losses are 1000000.0000 and 500200.0.
Notice that the log losses are just the mean of the batch losses since the start:
1000000 = 1e6 / 1
500200 = (1e6 + 400)/2

If you prefer to plot the actual batch losses rather than the mean losses (since the start of each epoch), then you can use the following callback. Of course, **this will only work with TF 2.2 or above**:

```python
class ExponentialLearningRate(keras.callbacks.Callback):
    def __init__(self, factor):
        self.factor = factor
        self.rates = []
        self.losses = []
    def on_epoch_begin(self, epoch, logs=None):
        self.prev_loss = 0
    def on_batch_end(self, batch, logs=None):
        batch_loss = logs["loss"] * (batch + 1) - self.prev_loss * batch
        self.prev_loss = logs["loss"]
        self.rates.append(K.get_value(self.model.optimizer.lr))
        self.losses.append(batch_loss)
        K.set_value(self.model.optimizer.lr, self.model.optimizer.lr * self.factor)
```

This works because `logs["loss"]` if the mean loss since the start of the epoch, so if you multiply it by the number of batches since the beginning of the epoch (which is `batch + 1`), then you get the sum of all batch losses since the start of the epoch. If you save the previous `logs["loss"]`, and you compute `prev_loss * batch`, you get the sum of all batch losses at the previous batch. The difference between the current sum of all losses and the previous sum of all losses is the current batch loss.

Hope this helps.

**Comment by kasrahabib** [comment]
Thanks!
